Cdas 69 mosaic xmlrpc client implementation

Lazer2077 opened the pull request:
<!-- Thanks for the contribution, this is awesome. -->

# PR Details
## Description
This is the implementation for the Carla XMLRPC Client 
<!--- Describe your changes in detail -->

## Related GitHub Issue

<!--- This project only accepts pull requests related to open GitHub issues or Jira Keys -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please DO NOT name partially fixed issues, instead open an issue specific to this fix -->
<!--- Please link to the issue here: -->

## Related Jira Key
CDAS-69 Ambassadors for MOSAIC Integration
<!-- e.g. CAR-123 -->

## Motivation and Context
The CARLA-MOSAIC integration currently uses a TraCI-based bridge for communication between CARLA and MOSAIC, managing actors (vehicles, pedestrians) and traffic lights. We have recently implemented an XML-RPC Server in CARLA to enable a more efficient, dedicated connection. To fully transition from the TraCI-based bridge, an XML-RPC Client must be implemented in the CARLA Ambassador within MOSAIC. This client will interact with the XML-RPC Server to handle bidirectional commands for actor lifecycle (create, update, destroy) and traffic light state updates, ensuring seamless synchronization between SUMO and CARLA simulations without reliance on TraCI.
<!--- Why is this change required? What problem does it solve? -->

## How Has This Been Tested?
Pass the docker build test
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Types of changes
Mainly change dir: 
1. cdasim/co-simulation/fed/mosaic-carla
2. cdasim/co-simulation/lib/mosaic-interactions

List of files been modified:
1. `CarlaAmbassador.java`: processing the interaction from rti
2. `pom.xml`: Add the build package
3. `CarlaXmlRpcClient.java`: register the corresponding Server function 

Related Interaction will be implemented and extend on the `rti` intereaction, list of file been added:

1. `CarlaActorRequest.java`
5. `CarlaActorResponse.java`
6. `CarlaTrafficLightRequest.java`
7. `CarlaTrafficLightResponse.java`
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)


## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have added any new packages to the sonar-scanner.properties file
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING**](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md) document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

Commit:
First version of Client

Commit:
Reconstruct SimulationStep.java Remove traci related implementation on CarlaAmbassador.java

Commit:
update license merge with the 66

Commit:
update license for Client

Commit:
delete unnecessary files

Commit:
update tls manager parameter interface

west053 committed:
update error

Commit:
CarlaAmbassador update

west053 committed:
update error

Commit:
change carla pom

Commit:
version 1.0 update, unrelated files deleted

chengyuan0124 committed:
Rename parameter

chengyuan0124 committed:
Added carlaActorLibRPCUrl

chengyuan0124 committed:
Update CarlaAmbassador.java

chengyuan0124 committed:
Set sumo as a tl manager

chengyuan0124 committed:
Added traffic for Town04

chengyuan0124 committed:
Added carlaActorLibRPCUrl

chengyuan0124 committed:
Set sensor as null for testing only carla and sumo

chengyuan0124 committed:
Merge branch 'develop' into cdas-69-mosaic_xmlrpc_client

Commit:
Multiple Connection for XMLRPC Lib  v 0.1.0

chengyuan0124 commented on co-simulation/fed/mosaic-carla/src/main/java/org/eclipse/mosaic/fed/carla/carlaconnect/CarlaMultiXmlRpcManager.java:
Do you think if it a good idea to remove all CARLA related functions from CarlaMutiXmlRpcManager.java, such as `createSensor`, `advanceSimulation`, `getAllActors` since they've been developed in CarlaXmlRpcClient.java already. In CarlaAmbassador.java, we can call `CarlaMultiXmlRpcManager.getClient(serverType).createSensor(...)`, for example, directly to use the function from CarlaXmlRpcManager or is there any obstacle for this approach?

Lazer2077 replied:
Ok, will do

west053 committed:
update error

Commit:
solved updated issues

Commit:
command record

west053 committed:
error

Commit:
issues solving test

west053 committed:
test image

Commit:
update carla config

west053 committed:
connection sucesfully

Commit:
Sumo ambassador update

west053 committed:
update error

Commit:
update Abs Sumo

west053 committed:
update error

Commit:
update

west053 committed:
update log

Commit:
update Anbassador for Vehicle management

Commit:
update logger

Commit:
update comment for the PR

Commit:
remove CarlaMultiXMLRPC management

Commit:
delete unrelated image

west053 committed:
update error

Commit:
update import

Commit:
Update the intereaction, try to use the MOSAIC official implementation

west053 committed:
compile error

Commit:
Update sumoAmbassador intereaction

west053 committed:
error

Commit:
update solving issues

west053 committed:
error

Commit:
solve error

Commit:
update log

chengyuan0124 commented on co-simulation/fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/ambassador/SumoAmbassador.java:
It appears you are considering adding a VehicleUpdates handler in the SUMO Ambassador. The SUMO Ambassador extends AbstractSumoAmbassador, which already implements a VehicleUpdates interaction handler [here](https://github.com/usdot-fhwa-stol/cdasim/blob/818a84eaf59edb2435b4a15100a0abab9b69dc30/co-simulation/fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/ambassador/AbstractSumoAmbassador.java#L557). This handler does not need to be modified or duplicated in the SUMO Ambassador, as it already fully supports vehicle synchronization for external vehicles. In [receiveInteraction(VehicleUpdates)](https://github.com/usdot-fhwa-stol/cdasim/blob/818a84eaf59edb2435b4a15100a0abab9b69dc30/co-simulation/fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/ambassador/AbstractSumoAmbassador.java#L557), the handler stores external vehicle state updates (e.g., position, speed, heading) in the externalVehicleMap [here](https://github.com/usdot-fhwa-stol/cdasim/blob/818a84eaf59edb2435b4a15100a0abab9b69dc30/co-simulation/fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/ambassador/AbstractSumoAmbassador.java#L566) if it is not existing in the externalVehicleMap and removes vehicles via TraCI if needed. These stored updates are then used in the processTimeAdvanceGrant method, specifically in [setExternalVehiclesToLatestPositions](https://github.com/usdot-fhwa-stol/cdasim/blob/818a84eaf59edb2435b4a15100a0abab9b69dc30/co-simulation/fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/ambassador/AbstractSumoAmbassador.java#L1234), where the stored position and heading data are applied to SUMO using TraCI’s moveToXY command during each simulation time step. This ensures external vehicle states are synchronized with SUMO. No changes are needed to either the SumoAmbassador or AbstractSumoAmbassador. Instead, ensure the CARLA Ambassador correctly triggers VehicleUpdates interactions with accurate vehicle status data, which the SUMO Ambassador will process automatically.

Lazer2077 replied:
OK, I delete the duplicated VehilceUpdates intereaction on the `sumoAmbassador.java`

chengyuan0124 replied:
I think it is the same for the traffic light. AbstractSumoAmbassador is using [TrafficLightStateChange](https://github.com/usdot-fhwa-stol/cdasim/blob/084b830141b8c126b45a3bde6c31d4e5c6e97361/co-simulation/fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/ambassador/AbstractSumoAmbassador.java#L773)

Lazer2077 replied:
OK, done

Commit:
update runtime, add the Trafficstate Change, delete duplicated vehicle update

west053 committed:
error

Commit:
resolve the spawn actor issues

Commit:
update the position offset

Commit:
Delete Traffic Light Update

Commit:
update v0.10 server

Commit:
adaptive server test

Commit:
update the transofrmation

Commit:
update transformation

Commit:
update input frame: sumo and carla

west053 committed:
update test

west053 committed:
Merge branch 'cdas-69-mosaic_xmlrpc_client' of https://github.com/usdot-fhwa-stol/cdasim into cdas-69-mosaic_xmlrpc_client

Commit:
print carla

Commit:
Merge branch 'cdas-69-mosaic_xmlrpc_client' of github.com:usdot-fhwa-stol/cdasim into cdas-69-mosaic_xmlrpc_client

Commit:
server print change

Commit:
add spectator_to_actor

west053 committed:
update test file

Commit:
Check avaliability

Commit:
Merge branch 'cdas-69-mosaic_xmlrpc_client' of github.com:usdot-fhwa-stol/cdasim into cdas-69-mosaic_xmlrpc_client

west053 committed:
update estimate range for town 4

Commit:
print extent x

Commit:
extent_x update

west053 committed:
update test

west053 committed:
Merge branch 'cdas-69-mosaic_xmlrpc_client' of https://github.com/usdot-fhwa-stol/cdasim into cdas-69-mosaic_xmlrpc_client

Commit:
udpate extent

Commit:
Merge branch 'cdas-69-mosaic_xmlrpc_client' of github.com:usdot-fhwa-stol/cdasim into cdas-69-mosaic_xmlrpc_client

Commit:
set offset

west053 committed:
update error

west053 committed:
Merge branch 'cdas-69-mosaic_xmlrpc_client' of https://github.com/usdot-fhwa-stol/cdasim into cdas-69-mosaic_xmlrpc_client

Commit:
update error solving

Commit:
offset update

Commit:
update logger

Commit:
update spawn

west053 committed:
UPDATE

west053 committed:
Merge branch 'cdas-69-mosaic_xmlrpc_client' of https://github.com/usdot-fhwa-stol/cdasim into cdas-69-mosaic_xmlrpc_client

Commit:
server don't need read net offset

Commit:
update server

west053 committed:
spawn overlap

west053 committed:
error

Commit:
add z shift and jetter

west053 committed:
example image

chengyuan0124 commented on co-simulation/bridge/net_offset_reader.py:
I think we move the offset reader to CARLA Ambassador, do we still need this?

Lazer2077 replied:
OK, I will implemented offset reader on CARLA Ambassador

chengyuan0124 replied:
I am not sure if we still need this or not. your reply mentioned it is moved to CARLA Ambassador but still having script here

chengyuan0124 commented on co-simulation/bridge/carla_xmlrpc_server.py:
what is this xml-rpc server, what is the differences between and 0.10 and 0.9

Lazer2077 replied:
I aim to integrate this together, currently I will delete and archive

chengyuan0124 commented on .vscode/settings.json:
Please don't push your vscode setting to git

Lazer2077 replied:
OK, sorry about that

chengyuan0124 commented on co-simulation/bridge/carla_v0.9_xmlrpc_server.py:
```diff
@@ -64,6 +68,21 @@ def __init__(self, host: str = 'localhost', port: int = 8090,
+        
+        # Try to read net offset from Town04.net.xml, fallback to hardcoded value
+      
+        self.net_offset_xy: Tuple[float, float] = (503.02, 423.76)
```
I am not sure we want to have offset here, for example, if we want to integrate other simulation in the future, this default will cause the problem since this is specific for SUMO

chengyuan0124 commented on co-simulation/bridge/carla_v0.9_xmlrpc_server.py:
```diff
@@ -64,6 +68,21 @@ def __init__(self, host: str = 'localhost', port: int = 8090,
+        # External-to-CARLA coordinate transform settings (SUMO/MOSAIC frame → CARLA frame)
+        # - input_frame: 'sumo' applies BridgeHelper-like conversion (y inversion, yaw - 90 deg, offset)
+        # - net_offset_xy: offset from SUMO net (x, y) applied before handedness flip
+        self.input_frame: str = 'sumo'
```
Does it mean it only supports 1 simulation in a run? Why CARLA XML-RPC server need to know about SUMO information?

Lazer2077 replied:
`input_frame` can be changed through the api, and it just a flag determine what should server handling the vehicle transformation

west053 committed:
Added converted sumo network file path to Town4 carla config

west053 committed:
Added private fields for tlLogic mappings

west053 committed:
Added carlaConfig sumoNetworkPath file handling

west053 committed:
Added parseSumoNetwork method

Commit:
add manual control

Commit:
Merge branch 'cdas-69-mosaic_xmlrpc_client' of github.com:usdot-fhwa-stol/cdasim into cdas-69-mosaic_xmlrpc_client

chengyuan0124 commented on co-simulation/bridge/carla_v0.10_xmlrpc_server.py:
Why we add manual_control.py here? manual_control.py is the script to check it is working, we don't really need to include it here

Lazer2077 replied:
You mean I added manual_control on the `cala_v0.10_xmlrpc_server.py`?

Commit:
chore: remove .vscode from repo

Commit:
delete .vscode

Commit:
Update Carla Ambassador and remove transformation function on server

Commit:
Remove manual control files from tracking and add to gitignore

Commit:
manual control, original implementation

Commit:
git ignore update

Commit:
update offset reader

Commit:
set default

west053 committed:
Implemented parseSumoNetwork method

west053 committed:
Added method for resolving tls state mask given a phase index, tlLogic, and program ID

west053 committed:
Added setCarlaTLColor method

west053 committed:
Added state mask char to color helper

west053 committed:
Added method to apply a state mask for a given tlLogic group

west053 committed:
Updated tls ChangePhase logic

west053 committed:
Added logic to apply timers to CARLA TLs

west053 committed:
Updated receiveInteraction for new TLS logic

chengyuan0124 committed:
Added sumoNetXmlPath in carla_config.json

chengyuan0124 committed:
Updated logic to read from config instead of env var

west053 committed:
Added sumo net xml file to CarlaConfiguration def

west053 committed:
Import fix

west053 committed:
Update to mosaic.sh to enable logging

Commit:
UPDATE MANUAL CONTROL

Commit:
both spawn support

Commit:
default point

west053 committed:
wrong collision

Commit:
get waypoint

Commit:
1

west053 committed:
2

Commit:
3

Commit:
update bi directional sync

Commit:
updated docker movement

west053 committed:
4

Commit:
5

Commit:
update Sync again

west053 committed:
mosaic.sh fix/logging support

west053 committed:
Update to mosaic.sh for logging

west053 committed:
Aligning sumo net path name with Carla

Commit:
reset

Commit:
Merge branch 'cdas-69-mosaic_xmlrpc_client' of github.com:usdot-fhwa-stol/cdasim into cdas-69-mosaic_xmlrpc_client

Commit:
Revert "5"

This reverts commit 920de8d6716f7d132cb15521cd4dd8f08641db30.

Commit:
update manual control vehicle removal

west053 committed:
Revert "reset"

This reverts commit 0eef4da4ca162fd1eec5935208f69eabc0f09a46.

west053 committed:
6

Commit:
7

Commit:
update logger

Commit:
update get all actor velocity

Commit:
Ambassador update

west053 committed:
9

Commit:
10

Commit:
Refactor CARLA actor data handling in CarlaAmbassador for improved synchronization with SUMO. Removed detailed logging of added/updated actors and replaced with streamlined conversion to VehicleData. Introduced a new method for converting CARLA actor information to VehicleData, enhancing clarity and maintainability. Updated logging to include speed during actor spawning and updates.

Commit:
update velocity

west053 committed:
File path fix

west053 committed:
Added parseSumoNetwork to CarlaAmbassador init

west053 committed:
Revert "update logger"

This reverts commit 779eb254c9b0f0bc0717e4f7d0bd1fb859ab6b91.

west053 committed:
Update CarlaAmb to receive TrafficLightUpdates interaction

west053 committed:
Update carla v0.9 XmlRpc set_tl_timer

west053 committed:
Added TrafficLightUpdates sub to CarlaAmb

west053 committed:
Debug statements

west053 committed:
Debug statements

west053 committed:
debug

west053 committed:
11

west053 committed:
Update logger

west053 committed:
Manual TrafficLightUpdates creation in sumo amb

west053 committed:
delete update speed

west053 committed:
Refactor CARLA XML-RPC client for improved connection handling and performance.

west053 committed:
compile sucessfully

west053 committed:
latest

west053 committed:
Added landmark id visualization script

chengyuan0124 committed:
Removed old client connection

west053 committed:
Carla xml rpc server odr id to carla id logic

west053 committed:
Added Carla TL freeze to server init

west053 committed:
remove old version bridge

west053 committed:
update transform

west053 committed:
Added carla to sumo state mapping method

west053 committed:
fix external vehicle problem

west053 committed:
Updated carla tl to sumo mapping

changed carla xml rpc server endpoint to send opendrive id

west053 committed:
Added call to  triggerInteraction for tl state change

west053 committed:
Testing Carla tl control

west053 committed:
update transformation

west053 committed:
update spectator location error

west053 committed:
Debug logging

west053 committed:
location sync

west053 committed:
Debug statements

west053 committed:
processTimeAdvanceGrant debugging

west053 committed:
Set isSimulationStep to true for processTimeAdvanceGrant

west053 committed:
Resolve freq spawn actor on sumo

kjrush commented on co-simulation/fed/mosaic-carla/src/main/java/org/eclipse/mosaic/fed/carla/carlaconnect/CarlaXmlRpcClient.java:
```diff
@@ -30,79 +31,1344 @@
+     * @param currentState Current actor state
+     * @return true if state has changed
+     */
+    private boolean hasActorStateChanged(Map<String, Object> previousState, Map<String, Object> currentState) {
```
Is there a reason for this being a Map<String, Object> instead of a proper class defition for `ActorState` or something similar? If the key-types here are truly heterogeneous then there's almost certainly a better way to represent this data other than as raw Object types. I see this pattern (use of raw Object) in a few places in this file, I'm assuming that's because we're getting untyped data from the XML-RPC interface (maybe I'll figure out more later in my review, I'll come back to this comment if so). Ideally in that case we should be converting those to proper types and classes at the earliest possible moment, throwing exceptions as appropriate if something happens incorrectly. Passing Object types around just subverts much of the type system's value and makes it more likely for us to introduce errors or maintenance difficulties.

west053 committed:
Exact location

west053 committed:
delete comments

west053 committed:
Added tl to odr mapping to xml rpc server

west053 committed:
Debug logging

west053 committed:
Updated getAllTrafficLightStates to handle Object[] deserialization

west053 committed:
Added isTlManager bool to Sumo/Carla Amb

kjrush commented on co-simulation/bridge/carla_v0.10_xmlrpc_server.py:
```diff
@@ -144,6 +190,92 @@ def _sim_timestamp(self) -> float:
+        try:
+            pitch_in = float(rotation_seq[0]) if len(rotation_seq) > 0 else 0.0
+            yaw_in = float(rotation_seq[1]) if len(rotation_seq) > 1 else 0.0
+            roll_in = float(rotation_seq[2]) if len(rotation_seq) > 2 else 0.0
```
Nitpick: Not generally a fan of using in-band values to represent out-of-band statuses. I.e. 0.0 is a valid pitch/yaw/roll, but based on the way this function is setup I can't distinguish "yaw not present" from "yaw is 0.0".

kjrush commented on co-simulation/bridge/carla_v0.10_xmlrpc_server.py:
```diff
@@ -144,6 +190,92 @@ def _sim_timestamp(self) -> float:
+
+    def _to_carla_rotation(self, rotation_seq: List[float]) -> carla.Rotation:
+        try:
+            pitch_in = float(rotation_seq[0]) if len(rotation_seq) > 0 else 0.0
```
This type conversion doesn't make a ton of sense to me, we're converting to float forcefully, but the type hint above already states the expected input is a List of floats, so `rotation_seq[i]` should be a float already no?

chengyuan0124 replied:
@Lazer2077

Lazer2077 replied:
_to_carla_rotation this function no longer used

kjrush commented on co-simulation/bridge/carla_v0.10_xmlrpc_server.py:
```diff
@@ -144,6 +190,92 @@ def _sim_timestamp(self) -> float:
+                    x_center = x_in - math.cos(math.radians(yaw)) * extent_x
+                    y_center = y_in - math.sin(math.radians(yaw)) * extent_x
+                    z_center = z_in - math.sin(math.radians(pitch_in)) * extent_x
+                except Exception:
```
What exceptions are we anticipating here? We're pretty aggressively checking for None and bounds checking the list accesses, I don't think these math operations can fail, is there some other concern?

EricChen-Lei replied:
It would be better to log down exception for debugging purposes

kjrush commented on co-simulation/bridge/carla_v0.10_xmlrpc_server.py:
```diff
@@ -83,6 +94,34 @@ def __init__(self, host: str = 'localhost', port: int = 8090,
+                        return actor
+                except Exception:
+                    continue
+        return None
```
Unclear if I haven't reached where this is used, but a failure to initialize an actor despite multiple attempts should be logged or at least printed to stdout/stderr.

kjrush commented on .cursor/debug.log:
Lets make sure to clean up any local workspace/IDE files before committing.

Lazer2077 replied:
done

kjrush commented on co-simulation/bridge/__pycache__/manual_control.cpython-311.pyc:
Why do we have a compiled binary file committed into the repo? Is this associated with a source file that needs to be reviewed?

kjrush commented on co-simulation/bridge/carla_v0.10_xmlrpc_server.py:
```diff
@@ -75,6 +77,15 @@ def __init__(self, host: str = 'localhost', port: int = 8090,
+        # External-to-CARLA coordinate transform settings (SUMO/MOSAIC frame → CARLA frame)
+        # - input_frame: 'sumo' applies BridgeHelper-like conversion (y inversion, yaw - 90 deg, offset)
+        # - net_offset_xy: offset from SUMO net (x, y) applied before handedness flip
+        self.input_frame: str = 'sumo'
```
Nitpick, not sure I love string as a type for this, but probably not worth changing. If there's only two possible values, a bool would make more sense `is_sumo_coordinate_frame` or something. Otherwise an enum type with entries for each possible coordinate system, or just directly taking in the specific coordinate transform parameters might be cleaner.

Strings are just easy to introduce errors with since there's very little assurance of correctness that the language itself can provide.

Lazer2077 replied:
deleted

kjrush commented on co-simulation/bridge/carla_v0.10_xmlrpc_server.py:
```diff
@@ -127,11 +166,18 @@ def _register_methods(self):
         self.server.register_function(self.load_map, 'load_map')
 
+        # Coordinate transform configuration
+        self.server.register_function(self.set_input_frame_mode, 'set_input_frame_mode')
```
How is this used in the ambassador side? Is there ever a case where the input frame isn't sumo?

Lazer2077 replied:
deleted

kjrush commented on co-simulation/bridge/carla_v0.10_xmlrpc_server.py:
```diff
@@ -144,6 +190,92 @@ def _sim_timestamp(self) -> float:
+            x_in = float(location_seq[0])
+            y_in = float(location_seq[1])
+            z_in = float(location_seq[2]) if len(location_seq) > 2 else 0.0
+        except Exception:
```
What exception are we anticipating here?

EricChen-Lei replied:
It would be better to log down exception for debugging purposes

kjrush commented on co-simulation/bridge/carla_v0.10_xmlrpc_server.py:
```diff
@@ -144,6 +190,92 @@ def _sim_timestamp(self) -> float:
+            return carla.Rotation(pitch_in, yaw_in - 90.0, roll_in)
+        return carla.Rotation(pitch_in, yaw_in, roll_in)
+
+    def _to_carla_velocity(self, velocity_seq_or_dict: Any) -> carla.Vector3D:
```
If it works, it works, but this function feels very strange from a type-system perspective. It implies we'd have a case where we're not sure whether or not the value we have is a dict or seq, but we know it contains velocity data. Which seems like a code smell to me that we may not be managing out types correctly

kjrush commented on co-simulation/bridge/carla_v0.10_xmlrpc_server.py:
```diff
@@ -144,6 +190,92 @@ def _sim_timestamp(self) -> float:
+    # ----- Coordinate transform configuration -----
+    def set_input_frame_mode(self, mode: str) -> bool:
+        try:
+            mode_l = str(mode).strip().lower()
```
As before, we've already type hinted this `mode` value as a str, why cast to str explicitly like this?

Lazer2077 replied:
deleted

kjrush commented on co-simulation/bridge/carla_v0.10_xmlrpc_server.py:
```diff
@@ -144,6 +190,92 @@ def _sim_timestamp(self) -> float:
+        try:
+            mode_l = str(mode).strip().lower()
+            if mode_l in ('sumo', 'carla'):
+                self.input_frame = 'sumo' if mode_l == 'sumo' else 'carla'
```
This ternary operator is redundant. Line 265 establishes that `mode_l` is either "sumo" or "carla" already. `self.input_frame = mode_l` achieves the same thing as the current ternary operator but without branching. Since if `mode_l` is "sumo", then it sets input frame to sumo, and if mode_l is carla it sets it to carla, and we've already ruled out that it could be anything else.

Lazer2077 replied:
deleted

kjrush commented on co-simulation/bridge/carla_v0.10_xmlrpc_server.py:
```diff
@@ -194,13 +326,36 @@ def connect(self) -> bool:
+                # Set CARLA simulation to passive mode (synchronous mode)
+                settings = self.world.get_settings()
+                settings.synchronous_mode = True
+                settings.fixed_delta_seconds = self.phase
```
Seeing how this value is used, I'm not sure phase is a good name for it. `delta_t` or `timestep_size` might be better or something.

Lazer2077 replied:
replaced

kjrush commented on co-simulation/bridge/carla_v0.10_xmlrpc_server.py:
```diff
@@ -194,13 +326,36 @@ def connect(self) -> bool:
                             CARLA_VERSION, self.carla_host, self.carla_port, current_map)
                 
+                # Automatically load Town04 if not already loaded
+                if current_map != "Town04":
```
I don't think we want to hard code loading Town04, we might run different scenarios on various CARLA maps.

Lazer2077 replied:
It is just temp solution for the Town04 test, I will implement this function later.

west053 committed:
advance_simulation

kjrush commented on co-simulation/bridge/carla_v0.10_xmlrpc_server.py:
```diff
@@ -418,14 +648,68 @@ def update_actor_velocity(self, actor_key: ActorKey, velocity: List[float]) -> b
     def get_all_actors(self) -> Dict[str, Dict[str, Any]]:
         try:
             with self.lock:
```
Maybe it'll be clear as I continue reading, but are we multithreading here? I haven't yet seen it. Is the XML RPC server itself handling the threading on a per-RPC basis?

Lazer2077 replied:
removed

kjrush commented on co-simulation/bridge/carla_v0.10_xmlrpc_server.py:
```diff
@@ -327,16 +482,66 @@ def spawn_actor(self, actor_type: str, actor_id: str,
+                        # Switch spectator to follow the first spawned actor with proper offset
+                        # Use the actual actor object for more reliable positioning
+                        self._set_spectator_to_actor_object(actor, 'follow', 8.0, 3.0, -20.0)
+                        print(f"========Spectator switched to follow first actor: {actor_id}========")
```
Would recommend converting all print statements to log statements using the logger object.

Lazer2077 replied:
replaced

kjrush commented on co-simulation/bridge/carla_v0.10_xmlrpc_server.py:
Most of my comments here can be interpreted as nitpicks if this has been thoroughly tested. The combination of type hints and lots of forced typecasting stands out to me as kind of odd and likely not the most maintainable way to implement this. Additionally, the prevalance of the

```
try:
expect Exception:
  continue
```

or similar pattern is a bit concerning, especially given that it surrounds a fair number of blocks that don't seem like they should throw exceptions. If they're throwing exceptions in practice, and this is a workaround, then we should probably address the underlying issue (likely caused by lack of None-checks or type checks somewhere further upstream in the stack).

EricChen-Lei commented on co-simulation/bridge/carla_v0.10_xmlrpc_server.py:
```diff
@@ -144,6 +190,92 @@ def _sim_timestamp(self) -> float:
+            pitch_in = float(rotation_seq[0]) if len(rotation_seq) > 0 else 0.0
+            yaw_in = float(rotation_seq[1]) if len(rotation_seq) > 1 else 0.0
+            roll_in = float(rotation_seq[2]) if len(rotation_seq) > 2 else 0.0
+        except Exception:
```
It would be better to log down exception for debugging purposes

Lazer2077 replied:
OK, I will log down these stuff on carla_v0.10_xmlrpc_server.py

west053 committed:
In TLUpdate interaction, use tlGroupInfo for to get state rather than custom mapping

west053 committed:
Carla TL freeze

west053 committed:
Set carla as tl manager

west053 committed:
Set sumo as tl manager

west053 committed:
Exit process TLStateChange in SumoAmb if not TL Manager

Enabled tl freeze in carla

west053 committed:
Added XML-RPC endpoint to freeze tls

west053 committed:
Moved carla tl freeze to first time advance

west053 committed:
Added freezeAllTLSumo method

west053 committed:
debugging statements cleanup

chengyuan0124 commented on co-simulation/bridge/carla_v0.9_xmlrpc_server.py:
```diff
@@ -116,11 +174,16 @@ def _register_methods(self):
         self.server.register_function(self.get_sensor_data, 'get_sensor_data')
```
For the actor xml-rpc server, we shouldn't include anything about sensor or is there any reason we do this?

Lazer2077 replied:
I will delete them

chengyuan0124 commented on co-simulation/bridge/carla_v0.10_xmlrpc_server.py:
I know during the testing, it is more focusing on v0.9, but do we do the parallel upgrade to v0.10 when some logic change in v0.9?

Lazer2077 replied:
It has been confirmed that the v0.10 has been updated

chengyuan0124 commented on co-simulation/bridge/__pycache__/manual_control.cpython-311.pyc:
Don't push the python compile file

Lazer2077 replied:
I will delete this stuff

chengyuan0124 commented on co-simulation/fed/mosaic-carla/src/main/java/org/eclipse/mosaic/fed/carla/ambassador/CarlaAmbassador.java:
```diff
@@ -112,6 +128,37 @@ public class CarlaAmbassador extends AbstractFederateAmbassador {
+     * SUMO net offset parsed from scenario .net.xml (x, y) in meters.
+     * Default to Town04 values if parsing fails.
+     */
+    private double[] sumoNetOffsetXY = new double[]{503.02, 423.76};
```
Do we need this default offset setting? Will this be overwritten by the value read from file?

Lazer2077 replied:
I'm not sure it will be overwritten or not; I can conduct test about that. I remember you implemented the reading stuff

chengyuan0124 replied:
No I didn't implement it, what I did was just adding parameter in carla_conifg.json in 2 of the pushes [[1]](https://github.com/usdot-fhwa-stol/cdasim/pull/252/commits/fc2f4b82d0691ffc5fcbfa1ba17553e399e9255f) [[2]](https://github.com/usdot-fhwa-stol/cdasim/pull/252/commits/1513889831e9c701dde7d5c79786092efaab83d3)

Lazer2077 replied:
It is confirmed that it is working after setting the offset to 0,0

chengyuan0124 commented on co-simulation/fed/mosaic-carla/src/main/java/org/eclipse/mosaic/fed/carla/ambassador/CarlaAmbassador.java:
```diff
@@ -211,101 +290,76 @@ public void initialize(long startTime, long endTime) throws InternalFederateExce
-                    log.warn("Error while connecting to CARLA simulator. Retrying.");
+                // Add sensor library server
+                // convert string null to null
+                if ("null".equalsIgnoreCase(carlaConfig.carlaSensorLibRPCUrl)) {
```
Not sure what is this line and line 307 do, if it is necessary to convert that from string to null, then we will have problem in line 297 since that is raw data without converting

Lazer2077 replied:
It just ensure that every format of Url can be processed

chengyuan0124 replied:
What I meant was that this logic is semantically backwards. Shouldn’t the string-to-null conversion happen before this `if (carlaConfig.carlaSensorLibRPCUrl != null || carlaConfig.carlaActorLibRPCUrl != null)` check? You’re trying to ensure the format is correct by converting `"null"` (as a string) into an actual `null`, but `carlaConfig.carlaSensorLibRPCUrl` and `carlaConfig.carlaActorLibRPCUrl` are being validated only after this condition runs. This makes the control flow confusing, because the if condition depends on values that haven’t been sanitized yet.

chengyuan0124 commented on co-simulation/fed/mosaic-carla/src/main/java/org/eclipse/mosaic/fed/carla/ambassador/CarlaAmbassador.java:
```diff
@@ -211,101 +290,76 @@ public void initialize(long startTime, long endTime) throws InternalFederateExce
+            } catch (MalformedURLException m) {
+                throw new InternalFederateException("Carla Ambassador initialization failed due to invalid XML-RPC server URLs! Check carla_config.json!");
             }
+        } else {
```
This `else` is pointless because these conditions are never true due to the logic of the parent `if`

Lazer2077 replied:
delete it

chengyuan0124 commented on co-simulation/fed/mosaic-carla/src/main/java/org/eclipse/mosaic/fed/carla/ambassador/CarlaAmbassador.java:
```diff
@@ -349,32 +402,232 @@ public synchronized void processTimeAdvanceGrant(long time) throws InternalFeder
+                if (multiXmlRpcManager != null) {
+                    multiXmlRpcManager.connectAll(60);
+                } else if (carlaXmlRpcClient != null) {
+                    carlaXmlRpcClient.connect(60);
```
What is carlaXmlRpcClient? is that for sensor xml rpc server? but I thought it will be saved into `multixmlrpcmanager`?

chengyuan0124 replied:
@Lazer2077

Lazer2077 replied:
I delete it

chengyuan0124 commented on co-simulation/fed/mosaic-carla/src/main/java/org/eclipse/mosaic/fed/carla/ambassador/CarlaAmbassador.java:
```diff
@@ -349,32 +402,232 @@ public synchronized void processTimeAdvanceGrant(long time) throws InternalFeder
+                if (multiXmlRpcManager != null) {
+                    tickClient = multiXmlRpcManager.getClient(CarlaXmlRpcClient.ServerType.ACTOR_LIB);
+                } else {
+                    tickClient = carlaXmlRpcClient;
```
same question, what is carlaXmlRpcClient? if it is sensor xml rpc, I don't recall it is calling tick

chengyuan0124 replied:
@Lazer2077

chengyuan0124 commented on co-simulation/fed/mosaic-carla/src/main/java/org/eclipse/mosaic/fed/carla/ambassador/CarlaAmbassador.java:
```diff
@@ -462,45 +720,405 @@ public boolean isTimeRegulating() {
+     * This replaces the external SimulationStep interaction with internal logic.
      */
-    public synchronized void triggerInteraction(int length, byte[] command) throws InternalFederateException {
+    private void triggerInternalSimulationStep() {
```
We can remove this function, this is not used anymore

Lazer2077 replied:
done

chengyuan0124 reviewed (changes requested):
Released my comment for xml-rpc server and CARLA Ambassador, will continue to review the reset

Commit:
review 1

Commit:
Remove .pyc files from repository and update .gitignore

Commit:
add the v0.10 server debug log

west053 committed:
update readme

west053 committed:
test delete

chengyuan0124 commented on co-simulation/fed/mosaic-carla/src/test/resources/carla_config.json:
Not sure what is this file, is this necessary?

chengyuan0124 commented on co-simulation/fed/mosaic-carla/src/test/resources/carla_config_debug.json:
Not sure what is this file, is this necessary?

chengyuan0124 commented on co-simulation/fed/mosaic-carla/src/test/resources/carla_config_with_map.json:
Not sure what is this file, is this necessary?

Lazer2077 replied:
deleted

chengyuan0124 commented on co-simulation/fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/ambassador/AbstractSumoAmbassador.java:
```diff
@@ -559,6 +560,43 @@ private synchronized void receiveInteraction(VehicleUpdates vehicleUpdates) thro
             return;
         }
 
+        // Handle added vehicles from external simulators (like CARLA)
```
Do not infer external vehicles from `VehicleUpdates`.
Instead, CARLA Ambassador should explicitly trigger a `VehicleFederateAssignment` firstly when a external vehicle created (e.x, in CARLA) before sending `VehicleUpdates` for that vehicle.”

chengyuan0124 replied:
did you update CARLA Ambassador to send `VehicleFederateAssignment` ? I only see you delete this section in `AbstractSumoAmbassador`

Lazer2077 replied:
updated

chengyuan0124 reviewed (changes requested):
Completed my initial review. Some logic in the CARLA–SUMO interaction needs to be adjusted. Specifically, when CARLA spawns a vehicle, it should trigger a `VehicleFederateAssignment` first, before sending any `VehicleUpdates`

chengyuan0124 commented on co-simulation/bridge/carla_v0.9_xmlrpc_server.py:
```diff
@@ -64,6 +70,12 @@ def __init__(self, host: str = 'localhost', port: int = 8090,
 
         self.lock = threading.RLock()
```
Why we are using lock in xmlrpc_server.py? It seems to me there is no multi-thread method in this python script

Lazer2077 replied:
what about the sensor lib? do we need lock to prevent conflict?

chengyuan0124 replied:
The carla_v0.9_xmlrpc_server.py is only for actor, it doesn't need to include the sensor support, sensor support function is in the carla-sensor-lib repo, two of them are different xml-rpc server. Is there any reason why you are considering the sensor in the actor-xml-rpc?

Lazer2077 replied:
OK, i delete the lock implementation

Commit:
delete uncessary json, delete infer external vehicles from VehicleUpdates on AbstractSumo

kjrush reviewed (commented):
Leaving partial comments, wont have time to finish this review within the sprint. I'll defer to @chengyuan0124 on whether/how these get resolve and if similar issues exist elsewhere, I've only had time to review this one file mostly.

west053 committed:
delete lock and opt server

west053 committed:
Add the VehicleFederateAssignment update

Commit:
update defalut map args

Commit:
update

west053 committed:
Removed sumo tl subscription debugging

west053 committed:
Removed comment

west053 committed:
name replaced

west053 committed:
Merge branch 'cdas-69-mosaic_xmlrpc_client' of https://github.com/usdot-fhwa-stol/cdasim into cdas-69-mosaic_xmlrpc_client

west053 committed:
handling the silent fallbacks

west053 committed:
Merge branch 'cdas-69-mosaic_xmlrpc_client' into cdas-80-veh_tls_sync

west053 committed:
Indentation fix

west053 committed:
Removed unimplemented endpoints

west053 committed:
Removed unused imports/fields

west053 committed:
Identation fix

Commit:
add ActorState Class

west053 committed:
Removed unimplemented endpoints

west053 committed:
Removed unused endpoints

west053 committed:
Moved tl id -> odr mapping to after scenario map load

Lazer2077 committed:
update for second review

Commit:
Merge branch 'cdas-69-mosaic_xmlrpc_client' of github.com:usdot-fhwa-stol/cdasim into cdas-69-mosaic_xmlrpc_client

chengyuan0124 committed:
Cdas 80 veh tls sync (#256)

<!-- Thanks for the contribution, this is awesome. -->

# PR Details
## Description
This PR introduces a CARLA XML-RPC–based control path and related
synchronization updates throught the CarlaAmbassador,
AbstractSumoAmbassador and the CARLA XML-RPC server to allow traffic
light sync through MOSAIC driven by CARLA or SUMO.

## Related GitHub Issue

<!--- This project only accepts pull requests related to open GitHub
issues or Jira Keys -->
<!--- If suggesting a new feature or change, please discuss it in an
issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps
to reproduce -->
<!--- Please DO NOT name partially fixed issues, instead open an issue
specific to this fix -->
<!--- Please link to the issue here: -->

## Related Jira Key
CDAS-80
## Motivation and Context
These changes allow traffic light synchronization to be handled entirely
through MOSAIC using XML-RPC.

## How Has This Been Tested?
These changes have been tested using the Town04 scenario and examined at
runtime that CARLA could emit TrafficLightStateChange interactions and
SUMO could receive and apply TrafficLightStateChange interactions. Also
verified SUMO could emit TrafficLightUpdates and CARLA side applied
corresponding XML-RPC state/timer calls and maintained frozen,
externally-driven sequencing.

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all
the boxes that apply: -->

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality
to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes
that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're
here to help! -->

- [ ] I have added any new packages to the sonar-scanner.properties file
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the
[**CONTRIBUTING**](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md)
document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

chengyuan0124 committed:
Merge branch 'develop' into cdas-69-mosaic_xmlrpc_client
